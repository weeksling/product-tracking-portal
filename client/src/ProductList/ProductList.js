import React, {Component} from 'react';

import { fetchProducts } from '../actions/productsActions';

import {Table} from 'material-ui/Table';

import { connect } from 'react-redux';


export class Product extends Component {
	render() {
		let {
			product
		} = this.props;

		return (
			<tr>
				<td>{product.product_id}</td>
				<td>{product.description}</td>
			</tr>
		)
	}
}

class ProductList extends Component {
	render() {
		let {
			products
		} = this.props;

		if (!products.length) {
			return (
				<p>No data</p>
			)
		}
		return (
			<div>
				<Table>
					{products.map( (product, index) =>{
						return <Product product={product} key={index} />
					})}
				</Table>
			</div>
		)
	}
}

export default ProductList;

