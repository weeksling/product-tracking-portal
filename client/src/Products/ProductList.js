import React, {Component} from 'react';

import { Link } from 'react-router-dom';


export const Loading = () => <p>Loading...</p>

export const ErrorMsg = () => <p>There's been an error</p>

export class Product extends Component {
	render() {
		let {
			product,
			selectProductToEdit
		} = this.props;

		return (
			<tr>
				<td>{product.product_id}</td>
				<td>{product.description}</td>
				<td><button onClick={ e => { selectProductToEdit(product) } }>edit</button></td>
				<td><button><Link to={"/"+product.product_id}>Track Locations</Link></button></td>
			</tr>
		)
	}
}

export default class ProductList extends Component {
	render() {
		let {
			loading,
			error,
			products,
			selectProductToEdit
		} = this.props;

		if (loading) {
	      return <Loading />
	    }

	    if (error) {
	      return <ErrorMsg />
	    }

		return (
			<div>
				<table>
					<thead>
						<th>Id</th>
						<th>Description</th>
					</thead>
					<tbody>
					{products.map( (product, index) =>{
						return <Product product={product} key={index} selectProductToEdit={selectProductToEdit} />
					})}
					</tbody>
				</table>
			</div>
		)
	}
}

