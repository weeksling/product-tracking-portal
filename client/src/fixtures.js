
export const mockProducts = [
    {
        "product_id": 1,
        "description": "Cesna 400"
    },
    {
        "product_id": 2,
        "description": "DC-6 Twin Beaver   sdss"
    },
    {
        "product_id": 3,
        "description": "Piper M602"
    }
]

export const mockLocations = [
    {
        "id": 1,
        "product": {
            "product_id": 1,
            "description": "Cesna 400"
        },
        "longitude": "-81.8149807",
        "latitude": "-81.8149807",
        "elevation": 444,
        "datetime": "2017-08-07T18:25:58Z"
    },
    {
        "id": 2,
        "product": {
            "product_id": 1,
            "description": "Cesna 400"
        },
        "longitude": "43.2583264",
        "latitude": "-81.8149807",
        "elevation": 500,
        "datetime": "2016-10-12T17:00:00Z"
    },
    {
        "id": 14,
        "product": {
            "product_id": 2,
            "description": "DC-6 Twin Beaver   sdss"
        },
        "longitude": "43.4591120",
        "latitude": "-80.3866930",
        "elevation": 500,
        "datetime": "2016-10-12T17:00:00Z"
    },
    {
        "id": 18,
        "product": {
            "product_id": 3,
            "description": "Piper M602"
        },
        "longitude": "44.4591120",
        "latitude": "-81.3866930",
        "elevation": 500,
        "datetime": "2016-10-15T17:00:00Z"
    },
    {
        "id": 19,
        "product": {
            "product_id": 3,
            "description": "Piper M602"
        },
        "longitude": "45.4591120",
        "latitude": "-82.3866930",
        "elevation": 600,
        "datetime": "2016-10-15T17:00:00Z"
    },
    {
        "id": 20,
        "product": {
            "product_id": 3,
            "description": "Piper M602"
        },
        "longitude": "46.4591120",
        "latitude": "-83.3866930",
        "elevation": 700,
        "datetime": "2016-10-15T17:00:00Z"
    }
]

export const mockOneProduct = {
    "product_id": 1,
    "description": "Cesna 400"
}

export const mockOneLocation = {
    "id": 1,
    "product": {
        "product_id": 1,
        "description": "Cesna 400"
    },
    "longitude": "-81.8149807",
    "latitude": "-81.8149807",
    "elevation": 444,
    "datetime": "2017-08-07T18:25:58Z"
}

