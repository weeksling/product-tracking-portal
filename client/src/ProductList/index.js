import React, {Component} from 'react';

import { fetchProducts } from '../actions/productsActions';

import ReactTable from 'react-table';

import { connect } from 'react-redux';

import ProductList from './ProductList';

export class ProductListContainer extends Component {
	componentDidMount() {
		this.props.dispatch(fetchProducts())
	}

	render() {
		let {
			loading,
			error,
			products
		} = this.props.products;

		return (
			<ProductList products = {products} />
		)
	}
}

export default connect( (store) => {
	return {
		products: store.products
	}
}) (ProductListContainer)

