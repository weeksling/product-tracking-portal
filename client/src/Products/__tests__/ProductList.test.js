import React from 'react';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import { mockProducts } from '../../fixtures.js';

import { MemoryRouter as Router } from 'react-router-dom';

import ProductList, { Product, ErrorMsg, Loading } from '../ProductList';

describe('ProductList', () => {

	const minProps = {
		products: [],
		loading: false,
		error: null
	}

	test('It doesn\'t blow up', () =>{
		expect(
			mount(<Router><ProductList {...minProps} /></Router>).exists()
		).toBe(true);
	})

	test('Displays a Loading element if loading is true', () => {
		expect(
			mount(<Router><ProductList {...minProps} loading={true} /></Router>).find(Loading).length
		).toBe(1);
	})

	test('Displays an Error if error is defined', () => {
		expect(
			mount(<Router><ProductList {...minProps} error={new Error()}/></Router>).find(ErrorMsg).length
		).toBe(1)
	})

	test('renders 3 Product elements if 3 products in props', ()=>{
		expect(
			mount(<Router><ProductList {...minProps} products={mockProducts} /></Router>).find(Product).length
		).toBe(3);
	})


	describe('Snapshots', () => {
		test('It renders properly with 3 products', () => {
			const tree = renderer.create(
			    <Router><ProductList {...minProps} products={mockProducts} /></Router>
			  ).toJSON();

			  expect(tree).toMatchSnapshot();
		})

		test('It renders properly with loading===true', () => {
			const tree = renderer.create(
			    <Router><ProductList {...minProps} loading={true} /></Router>
			  ).toJSON();
			
			  expect(tree).toMatchSnapshot();
		})

		test('It renders properly with error', () => {
			const tree = renderer.create(
			    <Router><ProductList {...minProps} error={new Error()} /></Router>
			  ).toJSON();
			
			  expect(tree).toMatchSnapshot();
		})
	})

})