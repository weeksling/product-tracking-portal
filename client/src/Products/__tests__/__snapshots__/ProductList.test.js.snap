// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList Snapshots It renders properly with 3 products 1`] = `
<div>
  <table>
    <thead>
      <th>
        Id
      </th>
      <th>
        Description
      </th>
    </thead>
    <tbody>
      <tr>
        <td>
          1
        </td>
        <td>
          Cesna 400
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            edit
          </button>
        </td>
        <td>
          <button>
            <a
              href="/1"
              onClick={[Function]}
            >
              Track Locations
            </a>
          </button>
        </td>
      </tr>
      <tr>
        <td>
          2
        </td>
        <td>
          DC-6 Twin Beaver   sdss
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            edit
          </button>
        </td>
        <td>
          <button>
            <a
              href="/2"
              onClick={[Function]}
            >
              Track Locations
            </a>
          </button>
        </td>
      </tr>
      <tr>
        <td>
          3
        </td>
        <td>
          Piper M602
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            edit
          </button>
        </td>
        <td>
          <button>
            <a
              href="/3"
              onClick={[Function]}
            >
              Track Locations
            </a>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`ProductList Snapshots It renders properly with error 1`] = `
<p>
  There's been an error
</p>
`;

exports[`ProductList Snapshots It renders properly with loading===true 1`] = `
<p>
  Loading...
</p>
`;
